MediaLibrary: MediaImageExporter and tests.

kurzee opened the pull request:
Continuing from #7117, this PR introduces the `MediaImageExporter` which does a few good bits of work.

Firstly, exporter errors are propagated upwards via `throw` as custom `enum` values so we can track exactly what went wrong in the process. The localized strings that result, however, should be generally written for a user to read. System errors are instead propagated as is via `MediaExportSystemError`.

The `ImageSourceWriter` utilizes CoreGraphics to do the heavy lifting of actually resizing images and writing them to disk. The comments _should_ help explain what's going on.

The goal was to consolidate how and where we process images for Media from a few places to just one. For example, the upcoming `MediaURLExporter` utilizes the `MediaImageExporter` to process images it detects as a URL, versus trying to do it itself. We'll also generate thumbnails of existing media via the same code in an upcoming implementation.

While there's a lot going here, the unit tests coverage should give us a nice handle on the implementation and ensuring the images are being processed as expected. For example, we've had long-running support issues with images not being resized as the user expects, based on the settings in app settings.

To test:
1. Make sure the code compiles.
2. Run the tests for `MediaImageExporterTests.swift`.

Needs review: @frosty and @SergioEstevao round three :surfer:

SergioEstevao commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -0,0 +1,257 @@
+    ///
+    func exportImage(_ image: UIImage, fileName: String?, onCompletion: @escaping (MediaImageExport) -> (), onError: @escaping (MediaExportError) -> ()) {
+        do {
+            guard let data = UIImageJPEGRepresentation(image, 1.0) else {
```
Are we sure we want maximum quality here? 
and shouldn't this be a parameter of the method? It could be added with a default value of 0.9

SergioEstevao commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -0,0 +1,257 @@
+        }
+    }
+
+    /// Configureable struct for writing an image to a URL from a CGImageSource, via CGImageDestination, particular to the needs of a MediaImageExporter.
```
Type on Configurable

SergioEstevao commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -0,0 +1,257 @@
+
+    /// Configureable struct for writing an image to a URL from a CGImageSource, via CGImageDestination, particular to the needs of a MediaImageExporter.
+    ///
+    /// - parameter url: File URL where the image should be written
```
Should this documentation be here? of on each parameter?

frosty commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -0,0 +1,257 @@
+        case imageSourceDestinationWriteFailed
+        var description: String {
+            switch self {
+            case .imageJPEGDataRepresentationFailed,
```
Are you checking every case here to make sure we don't miss any that may be added in the future?

kurzee replied:
I think... I kind of just kept putting them there 😄, we can use a default case for now though, but future adaptions would indeed want to catch any additions.

frosty commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -0,0 +1,257 @@
+                throw ExportError.imageSourceCreationWithDataFailed
+            }
+            guard let utType = CGImageSourceGetType(source) else {
+                throw ExportError.imageSourceIsAnUnknownType
```
I love this error handling at every step of the process!

frosty commented on WordPress/WordPressTest/MediaImageExporterTests.swift:
```diff
@@ -0,0 +1,366 @@
+    // MARK: - Image export testing
+
+    func testThatImageExportingByImageWorks() {
+        guard let mediaPath = OHPathForFile("test-image-device-photo-gps.jpg", type(of: self)) else {
```
Can we extract this into a helper method, as we use it in a lot of the tests?

frosty commented on WordPress/WordPressTest/MediaImageExporterTests.swift:
```diff
@@ -0,0 +1,366 @@
+                             fileName: nil,
+                             onCompletion: { (imageExport) in
+                                self.cleanUpExportedMedia(atURL: imageExport.url)
+                                expect.fulfill()
```
Do we need to check anything about the `imageExport` result before we fulfill?

kurzee replied:
Ah yes, two of these tests should indeed check for a valid image.

kurzee commented:
>And one question the images you are adding to the project? are they open source?

Good point, the GPS photos are my own photos so those are fine, need to check where this GIF came from though...

kurzee commented:
Replaced the gif with a gif created by @frosty himself, thanks @frosty!

Changed the image compression quality to default to `0.9` if an image requires a maximum size and will be resized. This way, users who prefer maximum size really are getting the maximum quality.

Ready for another look ⛷

frosty commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -136,6 +129,7 @@ class MediaImageExporter: MediaExporter {
             var writer = ImageSourceWriter(url: url, sourceUTType: type as CFString)
             if let maximumImageSize = maximumImageSize {
                 writer.maximumSize = maximumImageSize as CFNumber
+                writer.lossyCompressionQuality = 0.9 as CFNumber
```
Should we have a constant for `0.9` somewhere up top to make it slightly easier to find / tweak this in future?

SergioEstevao commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -117,24 +121,24 @@ class MediaImageExporter: MediaExporter {
     /// - parameter onError: Called if an error was encountered during creation.
     ///
-    func exportImageSource(_ source: CGImageSource, filename: String?, type: String, onCompletion: @escaping (MediaImageExport) -> (), onError: @escaping (MediaExportError) -> ()) {
+    func exportImageSource(_ source: CGImageSource, filename: String?, type: CFString, onCompletion: @escaping (MediaImageExport) -> (), onError: @escaping (MediaExportError) -> ()) {
```
Humm, not sure if having CFString has the parameter type here is the best option...

SergioEstevao commented on WordPress/Classes/Utility/Media/MediaImageExporter.swift:
```diff
@@ -117,24 +121,24 @@ class MediaImageExporter: MediaExporter {
             // Make a new URL within the local Media directory
             let url = try MediaLibrary.makeLocalMediaURL(withFilename: filename,
-                                                         fileExtension: fileExtensionForUTType(type),
+                                                         fileExtension: String.fileExtensionForUTType(type),
```
I see this more as an extension for FileManager or URL than String.

kurzee replied:
That's kind of the awkwardness of moving this into an extension, it really doesn't do much, and it doesn't rely on any particular class, such as FileManager.

kurzee replied:
I think I'll move to `URL` actually as it's expected to be working with a `URL` of sorts.
